jib-maven-plugin/src/main/java/com/google/cloud/tools/jib/maven/MavenLayerConfigurations.java

```diff
@@ -46,85 +43,74 @@
-
-            resourcesFiles.add(classFile);
-          });
-    }
+    Path classesOutputDirectory = Paths.get(project.getBuild().getOutputDirectory());
+
+    // Gets the classes files in the 'classes' output directory.
+    Predicate<Path> isClassFile = path -> path.toString().endsWith(".class");
+    addFilesToLayer(
+        classesOutputDirectory, isClassFile, classesExtractionPath, layerBuilder::addClassFile);
+
+    // Gets the resources files in the 'classes' output directory.
+    addFilesToLayer(
+        classesOutputDirectory,
+        isClassFile.negate(),
+        resourcesExtractionPath,
+        layerBuilder::addResourceFile);
 
     // Adds all the extra files.
     if (Files.exists(extraDirectory)) {
-      try (Stream<Path> extraFilesLayerDirectoryFiles = Files.list(extraDirectory)) {
-        extraFiles = extraFilesLayerDirectoryFiles.collect(Collectors.toList());
-
-      } catch (IOException ex) {
-        throw new IOException("Failed to list directory for extra files: " + extraDirectory, ex);
-      }
+      AbsoluteUnixPath extractionBase = AbsoluteUnixPath.get("/");
+      addFilesToLayer(extraDirectory, path -> true, extractionBase, layerBuilder::addExtraFile);
     }
 
-    AbsoluteUnixPath dependenciesExtractionPath =
-        appRoot.resolve(JavaEntrypointConstructor.DEFAULT_RELATIVE_DEPENDENCIES_PATH_ON_IMAGE);
-    AbsoluteUnixPath resourcesExtractionPath =
-        appRoot.resolve(JavaEntrypointConstructor.DEFAULT_RELATIVE_RESOURCES_PATH_ON_IMAGE);
-    AbsoluteUnixPath classesExtractionPath =
-        appRoot.resolve(JavaEntrypointConstructor.DEFAULT_RELATIVE_CLASSES_PATH_ON_IMAGE);
-
-    Builder layerBuilder = JavaLayerConfigurations.builder();
-    for (Path file : dependenciesFiles) {
-      layerBuilder.addDependencyFile(file, dependenciesExtractionPath.resolve(file.getFileName()));
-    }
-    for (Path file : snapshotDependenciesFiles) {
-      layerBuilder.addSnapshotDependencyFile(
-          file, dependenciesExtractionPath.resolve(file.getFileName()));
-    }
-    for (Path file : resourcesFiles) {
-      layerBuilder.addResourceFile(file, resourcesExtractionPath.resolve(file.getFileName()));
-    }
-    for (Path file : classesFiles) {
-      layerBuilder.addClassFile(file, classesExtractionPath.resolve(file.getFileName()));
-    }
-    for (Path file : extraFiles) {
-      layerBuilder.addExtraFile(file, AbsoluteUnixPath.get("/").resolve(file.getFileName()));
-    }
     return layerBuilder.build();
   }
 
+  @FunctionalInterface
+  private static interface FileToLayerAdder {
+
+    void add(Path sourcePath, AbsoluteUnixPath pathInContainer) throws IOException;
+  }
+
+  private static void addFilesToLayer(
+      Path sourceRoot,
+      Predicate<Path> pathFilter,
+      AbsoluteUnixPath basePathInContainer,
+      FileToLayerAdder addFileToLayer)
+      throws IOException {
+    Predicate<Path> isEmptyDirectory =
+        path -> Files.isDirectory(path) && path.toFile().list().length == 0;
```

Comment:
I think we should favor using `Files.list` since it at most would need to list one file from a directory rather than listing all the files every time.

Reply:
I used `File.list()` because `Files.list()` needs closing the stream. I can still go with `Files.list()`, but it doesn't seem to justify the reduced readability due to increased cognitive load. I think `path.toFile().list()` just works out fine here, since all we want to know is if it's empty.

Reply:
But it's also true that `toFile().list()` will construct the entire list, so `Files.list()` has an advantage too. Actually, let me use `Files.list()`.